### ISSUE TITLE
Autocomplete for aliases

### ISSUE THREAD

**Issue by farao** [fish-shell/fish-shell#6002]
It would be really nice to have autocomplete also when using aliases, e.g. when typing `git status<tab>` I get autocomplete, but when I have an alias gs for git status, I don't. Or is there already a way to achieve this?

I'm currently running fish 3.0.2.

**Comment by zanchey** [comment]
Aliases should set the `--wraps` option to the function they create. How have you created your aliases? If you used `alias gs="git show"; funcsave gs` a long time ago, you might need to recreate them.

**Comment by farao** [comment]
Oh, sorry, I just realized that the issue is a different: For two aliases, I had already existing programs in my PATH. So if there already exists a program with that name that the alias overwrites, then it tries to do the autocomplete for the original program and not for the alias.
E.g. I have an alias that is: `alias go="xdg-open"` BUT it uses the autocomplete pattern for the Go compiler ("/usr/bin/go") instead of the one of xdg-open - even though when executing `go <file>` xdg-open is used (as expected).

**Comment by faho** [comment]
>E.g. I have an alias that is: alias go="xdg-open"

I can't recommend doing that. Names are quite important in a shell, and overriding them for the sake of saving a few characters (instead of e.g. `open`, which we happen to set up by default) isn't worth it.

Anyway, what happens here is that the alias sets up "wrapping" completions in the function (like `function go --wraps xdg-open`), but still autoloads the completion file given by the original name.

Typically, this is what you want. E.g. if you set up `dnf` as a wrapper for yum, you still want the dnf completions to be read as well.

Fixing this would require some sort of flag to disable the original completions (e.g. `function go --only-wraps xdg-open`).

Again, I recommend picking a better name.

**Comment by zanchey** [comment]
> Typically, this is what you want

I'm not so sure about that - I alias (well, abbreviate actually, but the point still stands) `gs` to `git show` because I don't care about Ghostscript - and I certainly don't want its completions.

**Comment by farao** [comment]
I agree with zanchey, I think you never want the completions of another program that you're executing

**Comment by danberindei** [comment]
I'm using fish version 3.0.2 and my aliases are not created with `--wraps`:

``
> alias h "git help"
> functions gco
# Defined in - @ line 1
function gco --description 'alias gco git checkout'
	git checkout $argv;
end
```

I get proper autocomplete if I define the function manually to add `-w 'git checkout'.

**Comment by faho** [comment]
@danberindei Note that fish 3.0.2 didn't *show* "--wraps" definitions for functions, but it should be defined since 973533e37499dee6c816ae449cac9342be7e4dac.

If it isn't, the most likely issue is that you've overridden the "alias" function. Check `functions --details alias` to confirm where it is defined.

**Comment by danberindei** [comment]
@faho thanks, I see now that I did not get autocompletions because of https://github.com/fish-shell/fish-shell/issues/1625#issuecomment-528530953

**Comment by kaleb** [comment]
> Oh, sorry, I just realized that the issue is a different: For two aliases, I had already existing programs in my PATH. So if there already exists a program with that name that the alias overwrites, then it tries to do the autocomplete for the original program and not for the alias.
> E.g. I have an alias that is: `alias go="xdg-open"` BUT it uses the autocomplete pattern for the Go compiler ("/usr/bin/go") instead of the one of xdg-open - even though when executing `go <file>` xdg-open is used (as expected).

Perhaps an abbreviation would be better for this scenario using abbr.

**Comment by Croissong** [comment]
> I think you never want the completions of another program than you're executing

I agree with this as well. 
I have various aliases set for things like 
- `ls = eza`
- `find = fd`
- `cat = bat`

and I would assume others have these as well.  And I feel like abbreviations are not a good solution for these type of commands either. 
Although I am currently debating  - coming from zsh - whether to get rid of these renames altogether, given the fact that they can't be escaped as easily as in zsh (https://github.com/fish-shell/fish-shell/issues/5076)...

> Fixing this would require some sort of flag to disable the original completions (e.g. function go --only-wraps xdg-open).
This does sound like a good solution to me tho!
